src/mono/mono/mini/interp/intrinsics.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Intrinsics for libraries methods that are heavily used in interpreter relevant
+ * scenarios and where compiling these methods with the interpreter would have
+ * heavy performance impact.
+ */
+
+#include "intrinsics.h"
+
+#include <mono/metadata/object-internals.h>
+#include <mono/metadata/gc-internals.h>
+
+static guint32
+rotate_left (guint32 value, int offset)
+{
+        return (value << offset) | (value >> (32 - offset));
+}
+
+void
+interp_intrins_marvin_block (guint32 *pp0, guint32 *pp1)
+{
+	// Marvin.Block
+	guint32 p0 = *pp0;
+	guint32 p1 = *pp1;
+
+	p1 ^= p0;
+	p0 = rotate_left (p0, 20);
+
+	p0 += p1;
+	p1 = rotate_left (p1, 9);
+
+	p1 ^= p0;
+	p0 = rotate_left (p0, 27);
+
+	p0 += p1;
+	p1 = rotate_left (p1, 19);
+
+	*pp0 = p0;
+	*pp1 = p1;
+}
+
+guint32
+interp_intrins_ascii_chars_to_uppercase (guint32 value)
+{
+	// Utf16Utility.ConvertAllAsciiCharsInUInt32ToUppercase
+	guint32 lowerIndicator = value + 0x00800080 - 0x00610061;
+	guint32 upperIndicator = value + 0x00800080 - 0x007B007B;
+	guint32 combinedIndicator = (lowerIndicator ^ upperIndicator);
+	guint32 mask = (combinedIndicator & 0x00800080) >> 2;
+
+	return value ^ mask;
+}
+
+int
+interp_intrins_ordinal_ignore_case_ascii (guint32 valueA, guint32 valueB)
+{
+	// Utf16Utility.UInt32OrdinalIgnoreCaseAscii
+	guint32 differentBits = valueA ^ valueB;
+	guint32 lowerIndicator = valueA + 0x01000100 - 0x00410041;
+	guint32 upperIndicator = (valueA | 0x00200020u) + 0x00800080 - 0x007B007B;
+	guint32 combinedIndicator = lowerIndicator | upperIndicator;
+	return (((combinedIndicator >> 2) | ~0x00200020) & differentBits) == 0;
+}
+
+int
+interp_intrins_64ordinal_ignore_case_ascii (guint64 valueA, guint64 valueB)
+{
+	// Utf16Utility.UInt64OrdinalIgnoreCaseAscii
+	guint64 lowerIndicator = valueA + 0x0080008000800080l - 0x0041004100410041l;
+	guint64 upperIndicator = (valueA | 0x0020002000200020l) + 0x0100010001000100l - 0x007B007B007B007Bl;
+	guint64 combinedIndicator = (0x0080008000800080l & lowerIndicator & upperIndicator) >> 2;
+	return (valueA | combinedIndicator) == (valueB | combinedIndicator);
+}
+
+static int
+interp_intrins_count_digits (guint32 value)
+{
+	int digits = 1;
+	if (value >= 100000) {
+		value /= 100000;
+		digits += 5;
+	}
+	if (value < 10) {
+		// no-op
+	} else if (value < 100) {
+		digits++;
+	} else if (value < 1000) {
+		digits += 2;
+	} else if (value < 10000) {
+		digits += 3;
+	} else {
+		digits += 4;
+	}
+	return digits;
+}
+
+static guint32
+interp_intrins_math_divrem (guint32 a, guint32 b, guint32 *result)
+{
+	guint32 div = a / b;
+	*result = a - (div * b);
+	return div;
+}
```

Comment:
can be just
```c
*result = a % b;
return a / b;
```
C/C++ compilers are able to recognize a single div there https://godbolt.org/z/-7jsRb